Adding Coconut-SVSM support

ramagali24 opened the pull request:
Coconut-SVSM support leverages AMD-ES/AMD-SEV for setting up the host, launching the guest VM (by directly setting up snpguest within the guest VM), and performing attestation and measurement verification (using igvmmeasure versus snpguest).

larrydewey commented on tools/snp.sh:
```diff
@@ -90,12 +99,14 @@ QEMU_CMDLINE_FILE="${QEMU_CMDLINE:-${LAUNCH_WORKING_DIR}/qemu.cmdline}"
 GENERATED_INITRD_BIN="${SETUP_WORKING_DIR}/initrd.img"
 
-# URLs and repos
+# URLs and repois
```
```suggestion
# URLs and repos
```

larrydewey commented on tools/snp.sh:
```diff
@@ -90,12 +99,14 @@ QEMU_CMDLINE_FILE="${QEMU_CMDLINE:-${LAUNCH_WORKING_DIR}/qemu.cmdline}"
+AMDSEV_SVSM_BRANCH="svsm-latest"
 SNPGUEST_URL="https://github.com/virtee/snpguest.git"
-SNPGUEST_BRANCH="tags/v0.7.1"
+SNPGUEST_BRANCH="tags/v0.3.2"
```
You need to rebase off of upstream. This should be 0.7.1

larrydewey commented on tools/snp.sh:
```diff
@@ -268,12 +280,6 @@ install_dependencies() {
-  sudo apt install -y cpuid
-
-  # Needed to build 6.11.0-rc3 SNP kernel on the host
-  pip install tomli
```
Why are we removing the installation of critical components from prior builds?

larrydewey commented on tools/snp.sh:
```diff
@@ -515,19 +521,32 @@ get_guest_kernel_version() {
 
 # Save binary paths in source file
-cat > "${SETUP_WORKING_DIR}/source-bins" <<EOF
+if [ "$SVSM" = true ]; then
+    cat > "${SETUP_WORKING_DIR}/source-bins" <<EOF
 QEMU_BIN="${SETUP_WORKING_DIR}/AMDSEV/qemu/build/qemu-system-x86_64"
 OVMF_BIN="${SETUP_WORKING_DIR}/AMDSEV/ovmf/Build/AmdSev/DEBUG_GCC5/FV/OVMF.fd"
 INITRD_BIN="${GENERATED_INITRD_BIN}"
 KERNEL_BIN="${guest_kernel}"
 EOF
+else
+    cat > "${SETUP_WORKING_DIR}/source-bins" <<EOF
+QEMU_BIN="${SETUP_WORKING_DIR}/AMDSEV/qemu/build/qemu-system-x86_64"
+OVMF_BIN="${SETUP_WORKING_DIR}/AMDSEV/ovmf/Build/OvmfX64/DEBUG_GCC5/FV/OVMF.fd"
+INITRD_BIN="${GENERATED_INITRD_BIN}"
+KERNEL_BIN="${guest_kernel}"
+IGVM_FILE="${SETUP_WORKING_DIR}/AMDSEV/svsm/bin/coconut-qemu.igvm"
+EOF
+fi
```
Where the two logical statements are doing almost identical things for the first 4/5 of the condition, it make sense to remove most of it from the conditional statement and only append the IGVM_FILE code to the end of it when it is present. Keep in mind I have not tested the following code, so it may need some massaging.

```suggestion
cat > "${SETUP_WORKING_DIR}/source-bins" <<EOF
QEMU_BIN="${SETUP_WORKING_DIR}/AMDSEV/qemu/build/qemu-system-x86_64"
OVMF_BIN="${SETUP_WORKING_DIR}/AMDSEV/ovmf/Build/AmdSev/DEBUG_GCC5/FV/OVMF.fd"
INITRD_BIN="${GENERATED_INITRD_BIN}"
KERNEL_BIN="${guest_kernel}"
EOF
if [ -z "$SVSM" ]
then
    cat >> "${SETUP_WORKING_DIR}/source-bins" <<EOF
IGVM_FILE="${SETUP_WORKING_DIR}/AMDSEV/svsm/bin/coconut-qemu.igvm"
EOF
fi
```

ramagali24 replied:
Our main goal in adding SVSM is to avoid breaking the existing codebase. You can see differences in the OVMF path AMDSEV/ovmf/Build/OvmfX64/DEBUG_GCC5/FV/OVMF.fd, as well as with IGVM. Once Coconut starts using the upstream EDK2, this path will be overwritten. Your suggested changes are more extensive than what we currently have. Does it make sense to proceed with these changes?

larrydewey replied:
My point is, if you have only two changes, you only need to adjust those two things. You don't need to have all the rest of the stuff in your conditional statement. Yes, you need to fix this

larrydewey commented on tools/snp.sh:
```diff
@@ -542,19 +561,35 @@ copy_launch_binaries() {
-  cp "${OVMF_BIN}" "${LAUNCH_WORKING_DIR}"
-  #cp "${INITRD_BIN}" "${LAUNCH_WORKING_DIR}"
-  cp "${KERNEL_BIN}" "${LAUNCH_WORKING_DIR}"
+  echo " ${OVMF_BIN} ${IGVM_FILE}"
```
We probably don't need a debug line here.

larrydewey commented on tools/snp.sh:
```diff
@@ -542,19 +561,35 @@ copy_launch_binaries() {
+	  cp "${KERNEL_BIN}" "${LAUNCH_WORKING_DIR}"
+	  cp "${IGVM_FILE}" "${LAUNCH_WORKING_DIR}"
+	  
+  fi
```
Same here. If the kernel bin is going to be the same, we don't need to copy it into both conditionals. We should refactor this.

larrydewey commented on tools/snp.sh:
```diff
@@ -542,19 +561,35 @@ copy_launch_binaries() {
+KERNEL_BIN="${LAUNCH_WORKING_DIR}/$(basename "${KERNEL_BIN}")"
+EOF
+fi
 }
```
Same here. We shouldn't be duplicating work

larrydewey commented on tools/snp.sh:
```diff
@@ -585,7 +620,7 @@ build_base_qemu_cmdline() {
   add_qemu_cmdline_opts "-m ${GUEST_MEM_SIZE_MB}M"
   add_qemu_cmdline_opts "-no-reboot"
-  add_qemu_cmdline_opts "-vga none"
+  add_qemu_cmdline_opts "-vga std"
```
Would it make more sense to add a flag here if we want VGA enabled/disabled? @ryansavino thoughts?

larrydewey commented on tools/snp.sh:
```diff
@@ -594,7 +629,7 @@ build_base_qemu_cmdline() {
 
   # Storage
-  add_qemu_cmdline_opts "-device virtio-scsi-pci,id=scsi0,disable-legacy=on,iommu_platform=true,romfile="
+  add_qemu_cmdline_opts "-device virtio-scsi-pci,id=scsi0,disable-legacy=on,iommu_platform=true"
```
You need to rebase off upstream code. This will override the new `romfile` requirement for the devices.

larrydewey replied:
Still need to address this

larrydewey commented on tools/snp.sh:
```diff
@@ -646,11 +681,23 @@ set_acl_for_sev_device() {
   local amdsev_branch="${1:-${AMDSEV_DEFAULT_BRANCH}}"
-
+  
+    if [ "$SVSM" = true ]; then
+       local amdsev_branch="${1:-${AMDSEV_DEFAULT_BRANCH}}"
+	   
+    else
+       AMDSEV_URL=${AMDSEV_SVSM_URL}
+       amdsev_branch="svsm-latest"
+	   # Create directory
+       
+    fi	
```
Your logic seems backward here?

DGonzalezVillal replied:
@ramagali24 

I keep seeing in your code that you check to see if $SVSM is true, but then if it is true, you proceed to do the non-svsm process, is there a reason you are doing it this way?

DGonzalezVillal replied:
for example lines 528 -> 543

ramagali24 replied:
Addressed redundant re-declaration of amdsev_branch inside the if block.

larrydewey commented on tools/snp.sh:
```diff
@@ -664,17 +711,21 @@ build_and_install_amdsev() {
   git fetch current "${amdsev_branch}"
   git checkout "current/${amdsev_branch}"
-
+   
+  if [ "$SVSM" = true ]; then
   # Based on latest AMDSEV documentation
   # Delete the ovmf/ directory prior to the build step for ovmf re-initialization
   [ ! -d "ovmf" ] || rm -rf "ovmf"
+  fi
```
What is the purpose of this being wrapped in a conditional for SVSM only?

DGonzalezVillal replied:
@ramagali24 I'm thinking you mean to do this if SVSM flag was not passed? 

I think you have your conditionals mixed up

ramagali24 replied:
Your reasoning makes sense. Since SVSM is currently using an older OVMF build, removing the directory for a fresh build can potentially cause issues, especially if SVSM has dependencies on its own specific OVMF structure or build process. It's a good idea to address this later when both SNP and SVSM converge on the same OVMF version from upstream sources.

For now, to avoid disrupting the SVSM build, we can skip removing the directory when SVSM is involved.

larrydewey replied:
No. You can't skip this for the sake of speeding up the merging of your PR. Fixing the logic for this does nothing to break any level of the build process. It literally is centered around a conditional change. If you don't do clean up, and you were to build the SVSM then build the standard image, it will break the image as it doesn't clean up the SVSM from the previous build. You need to address it and fix it.

larrydewey commented on tools/snp.sh:
```diff
@@ -664,17 +711,21 @@ build_and_install_amdsev() {
   
   # Get guest kernel version from the guest config file
   local guest_kernel_version=$(get_guest_kernel_version)
+  
+  
```
Unnecessary spacing. Please remove

larrydewey commented on tools/snp.sh:
```diff
@@ -684,11 +735,13 @@ build_and_install_amdsev() {
   # so, copying bzImage file into guest kernel binary file if vmlinuz is absent inside the guest directory
   [ -f ${guest_kernel_bin} ] || cp -v ${bzImage_file} ${guest_kernel_bin}
-
+  
```
Unnecessary spacing. Please remove

larrydewey commented on tools/snp.sh:
```diff
@@ -684,11 +735,13 @@ build_and_install_amdsev() {
   cd $(ls -dt */ | grep snp-release- | head -1)
   sudo ./install.sh
   
+  
+  
```
Unnecessary spacing. Please remove

larrydewey commented on tools/snp.sh:
```diff
@@ -786,14 +839,19 @@ setup_and_launch_guest() {
 
   # snp object and kernel-hashes on
-  add_qemu_cmdline_opts "-object sev-snp-guest,id=sev0,cbitpos=51,reduced-phys-bits=1,kernel-hashes=on"
+  if [ "$SVSM" = true ]; then
+     add_qemu_cmdline_opts "-object sev-snp-guest,id=sev0,cbitpos=51,reduced-phys-bits=1,kernel-hashes=on"
+  else
+     add_qemu_cmdline_opts "-object memory-backend-memfd,size=2G,id=mem0,share=true,prealloc=false,reserve=off"
+     add_qemu_cmdline_opts "-object sev-snp-guest,id=sev0,cbitpos=51,reduced-phys-bits=1,init-flags=5,igvm-file="$IGVM_FILE""
+  fi
```
Again, your logic seems reversed here?

ramagali24 replied:
In what context is this applicable? Could you explain further? The else part is for SVSM, and we need to understand how to correctly configure the QEMU command for SVSM. There is slight differences in snp and svsm qemu commands.

larrydewey replied:
Your logic for the flag is backward. If SVSM is enabled, then do things for SVSM. If it is disabled, do things for the standard flow. Your logic is backward. That is what I mean.

larrydewey commented on tools/snp.sh:
```diff
@@ -786,14 +839,19 @@ setup_and_launch_guest() {
   add_qemu_cmdline_opts "-kernel ${KERNEL_BIN}"
   add_qemu_cmdline_opts "-append \"${GUEST_KERNEL_APPEND}\""
-
+   
```
Unnecessary spacing. Please remove

larrydewey commented on tools/snp.sh:
```diff
@@ -929,99 +987,22 @@ setup_guest_attestation() {
-        echo "turin"
-      fi
+    8*|9*)
+      cpu_code_name="genoa"
+      echo $cpu_code_name
       ;;
     *)
-      echo "Invalid CPU"
+      >&2 echo -e "Unknown CPU Model: ${cpu_model}"
+      return 1
```
Please rebase. You are missing Turin support

larrydewey commented on tools/snp.sh:
```diff
@@ -1081,42 +1062,134 @@ attest_guest() {
   # Request and display the snp attestation report with random data
-  ssh_guest_command "sudo ./snpguest report attestation-report.bin request-data.txt --random"
-  ssh_guest_command "./snpguest display report attestation-report.bin"
+  if [ "$SVSM" = true ]; then
+	ssh_guest_command "sudo ./snpguest report attestation-report.bin request-data.txt --random"
+  else
+	ssh_guest_command "sudo ./$Home/snpguest/target/release/snpguest report attestation-report.bin request-data.txt --random --vmpl 3"
+  fi
+  ssh_guest_command "./$Home/snpguest/target/release/snpguest display report attestation-report.bin"
```
Logic is backward here, again.

larrydewey commented on tools/snp.sh:
```diff
@@ -1081,42 +1062,134 @@ attest_guest() {
   # Retrieve ark, ask, vcek (saved in ./certs)
-  ssh_guest_command "./snpguest fetch ca pem ${cpu_code_name} . --endorser vcek"
-  ssh_guest_command "./snpguest fetch vcek pem ${cpu_code_name} . attestation-report.bin"
+  ssh_guest_command "./$Home/snpguest/target/release/snpguest fetch ca pem ${cpu_code_name} ."
+  ssh_guest_command "./$Home/snpguest/target/release/snpguest fetch vcek pem ${cpu_code_name} . attestation-report.bin"
```
This seems like some very different behavior from what we were doing previously. Is there a reason we are building the tool within the guest instead of on the host machine and copying it over? @ryansavino  thoughts?

ramagali24 replied:
Got it. So, installing and running snpguest directly in the guest VM ensures compatibility with the glibc and tools available on that VM, avoiding issues caused by differences between the host(ubuntu 24.04) and guest (ubuntu 22.02) environments. If you need any further adjustments or have more questions, just let me know.

larrydewey commented on tools/snp.sh:
```diff
@@ -1081,42 +1062,134 @@ attest_guest() {
-  ssh_guest_command "./snpguest verify attestation . attestation-report.bin"
+  ssh_guest_command "./$Home/snpguest/target/release/snpguest verify attestation . attestation-report.bin"
+  
+ if [ "$SVSM" = true ]; then 
+  
+	  # Use sev-snp-measure utility to calculate the expected measurement
+	  local expected_measurement=$(generate_snp_expected_measurement)
+	  echo -e "\nExpected Measurement (sev-snp-measure):  ${expected_measurement}"
+
+	  # Parse the measurement out of the snp report
+	  local snpguest_report_measurement=$(ssh_guest_command \
+		"./snpguest display report attestation-report.bin \
+		| tr '\n' ' ' \
+		| sed \"s|.*Measurement:\(.*\)Host Data.*|\1\n|g\" \
+		| sed \"s| ||g\"")
+
+	  # Remove any special characters and print the value
+	  snpguest_report_measurement=$(echo ${snpguest_report_measurement} | sed $'s/[^[:print:]\t]//g')
+	  echo -e "Measurement from SNP Attestation Report: ${snpguest_report_measurement}\n"
+
+	  # Compare the expected measurement to the guest report measurement
+	  [[ "${expected_measurement}" == "${snpguest_report_measurement}" ]] \
+		&& echo -e "The expected measurement matches the snp guest report measurement!" \
+		|| { >&2 echo -e "FAIL: measurements do not match"; return 1; }
+else 
+ 	 Verify_measurement
+fi
+}
```
Your logic seems backward here. Perhaps check if the variable is set instead of true?

larrydewey commented on tools/snp.sh:
```diff
@@ -1081,42 +1062,134 @@ attest_guest() {
-
-  # Remove any special characters and print the value
-  snpguest_report_measurement=$(echo ${snpguest_report_measurement} | sed $'s/[^[:print:]\t]//g')
+  scp_guest_command "${GUEST_USER}@localhost:$HOME/attestation-report.bin" "$SETUP_WORKING_DIR/AMDSEV/svsm"
```
We should absolutely 100% not be doing any level of attestation on the Host Platform... This violates the trust boundary

DGonzalezVillal replied:
Plus we also already have logic to verify the measurement against an existing attestation report. This code is doing things that we already have in other parts of the script. You really only needed to add how to calculate the expected measurement for svsm.

larrydewey commented on tools/snp.sh:
```diff
@@ -1081,42 +1062,134 @@ attest_guest() {
-  # Use sev-snp-measure utility to calculate the expected measurement
-  local expected_measurement=$(generate_snp_expected_measurement)
-  echo -e "\nExpected Measurement (sev-snp-measure):  ${expected_measurement}"
+Verify_measurement(){
```
Why do we have two separate verification functions? Wouldn't it make sense to just update the existing one?

ramagali24 replied:
I am comparing with igvmmeasure and snpguest report. SNP does with sev-snpmeasure and snpguest report.

DGonzalezVillal replied:
For this one he is using the igvmmeasure rather than the sev-snp-measure tool. You will need a separate function to calculate the svsm expected measurement, but then maybe put that function back into the original verify measurement.

larrydewey replied:
Exactly what @DGonzalezVillal is saying. That, or put a flag in the function for verify, specifying you want to use SVSM logic. We already have the flag. It wouldn't be hard to do a check and adjust the logic.

larrydewey commented on tools/snp.sh:
```diff
@@ -1081,42 +1062,134 @@ attest_guest() {
+  # Assign the output of the hexdump command to a variable
+  snpguest_report_measurement=$(hexdump -s 0x90 -n 48 -v -e '48/1 "%02X"' attestation-report.bin)
+
+  # Print the variable to verify its content
+  echo "$snpguest_report_measurement"
```
Do we need to be outputting this value? We are doing it on the subsequent line...

larrydewey commented on tools/snp.sh:
```diff
@@ -1187,6 +1264,11 @@ main() {
   if ! $UPM; then
     SETUP_WORKING_DIR="${SETUP_WORKING_DIR}/non-upm"
   fi
+  
+  # Set SETUP_WORKING_DIR for non-upm
+  #if [ "$SVSM" = true ]; then
+   # SVSM_SETUP_WORKING_DIR="${SETUP_WORKING_DIR}/svsm"
+  #fi
```
Do we intend to support non-UPM?

DGonzalezVillal replied:
Idk if there are svsm kernels that support non-upm

larrydewey commented on tools/snp.sh:
```diff
@@ -1196,6 +1278,7 @@ main() {
       ;;
 
     setup-host)
+	  
```
Unnecessary spacing. Please remove

larrydewey commented on tools/snp.sh:
```diff
@@ -1205,6 +1288,7 @@ main() {
       fi
 
       source "${SETUP_WORKING_DIR}/source-bins"
+	  
```
Unnecessary spacing. Please remove

larrydewey commented on tools/snp.sh:
```diff
@@ -1223,8 +1307,11 @@ main() {
       # Changes in AMDESE/linux set debug_swap on by default and affect the measurement
-      sudo modprobe -r kvm_amd
-      sudo modprobe kvm_amd debug_swap=0
+	  if [ "$SVSM" = true ]; then
+		sudo modprobe -r kvm_amd
+        sudo modprobe kvm_amd debug_swap=0
+	  fi
+     
```
Any reason why we introduced a conditional around this? Seems odd to me...

ramagali24 replied:
Simple answer is this does not need for coconut-svsm. Without this works for me. I added svsm logic, based on scripts i shared though previous PR.

larrydewey replied:
This has nothing to do with coconut SVSM. This is a kernel module adjustment to make sure that the kernel feature for DEBUG_SWAP is disabled. Regardless of being inside of a COCONUT_SVSM machine or a standard guest VM running SEV-SNP, if this kernel module is enabled, it dramatically fills the buffer for the DMESG log and can potentially cause the guest to crash. This condition needs to be removed to make sure the guest is not inhibited.

larrydewey commented on tools/snp.sh:
```diff
@@ -1235,12 +1322,19 @@ main() {
     attest-guest)
+	if [ "$SVSM" = true ]; then
       install_rust
       install_sev_snp_measure
       install_dependencies
       wait_and_retry_command verify_snp_guest
-      setup_guest_attestation
-      attest_guest
+	  setup_guest_attestation
+	  attest_guest
+	else
+	  wait_and_retry_command verify_snp_guest
+	  setup_svsm_guest_attestation
+	  attest_guest
+    fi
+     attest_guest  	  
```
Indentation here seems really off. Are you using tabs vs spaces? Also, if we are repeating all but one line, we should adjust this conditional logic.

larrydewey reviewed (changes requested):
There are a number of points needing adjustment. The biggest one is this PR needs to be re-based on upstream. Please adjust and re-request review.

ramagali24 committed:
Coconut-SVSM support to invoke AMD-ES/AMD-SEV for setting up the host, launching the guest (directly setting up snpguest in the guest VM), and performing attestation & measurement verification (using igvmmeasure vs snpguest)

DGonzalezVillal commented on tools/snp.sh:
```diff
@@ -550,11 +563,19 @@ copy_launch_binaries() {
   cp "${KERNEL_BIN}" "${LAUNCH_WORKING_DIR}"
 
 # Save binary paths in source file
+if [ "$SVSM" = true ]; then
 cat > "${LAUNCH_WORKING_DIR}/source-bins" <<EOF
 OVMF_BIN="${LAUNCH_WORKING_DIR}/$(basename "${OVMF_BIN}")"
 INITRD_BIN="${LAUNCH_WORKING_DIR}/$(basename "${INITRD_BIN}")"
 KERNEL_BIN="${LAUNCH_WORKING_DIR}/$(basename "${KERNEL_BIN}")"
 EOF
+else  
+cat > "${LAUNCH_WORKING_DIR}/source-bins" <<EOF
+IGVM_FILE="${LAUNCH_WORKING_DIR}/$(basename "${IGVM_FILE}")"
+INITRD_BIN="${LAUNCH_WORKING_DIR}/$(basename "${INITRD_BIN}")"
+KERNEL_BIN="${LAUNCH_WORKING_DIR}/$(basename "${KERNEL_BIN}")"
+EOF
+fi
```
Consolidate logic like Larry mentioned in the file set-up

DGonzalezVillal commented on tools/snp.sh:
```diff
@@ -1107,16 +1152,102 @@ attest_guest() {
+ 	 Verify_measurement
+fi
+}
+setup_svsm_guest_attestation() {
+
+   ssh_guest_command "
+   
+    # Update and install necessary packages
+    echo 'Updating package list...'
+    sudo apt-get update
+
+    echo 'Installing necessary packages...'
+    sudo apt-get install -y git build-essential libtss2-dev tpm2-tools
+
+    echo 'Installing Rust...'
+    source ""${HOME}"/.cargo/env" 2>/dev/null || true
+
+    curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs -sSf | sh -s -- -y
+    source ""${HOME}"/.cargo/env" 2>/dev/null
+    "    
+
+  # Check if 'snpguest' directory exists and remove it if it does
+  ssh_guest_command "
+  if [ -d \"snpguest\" ]; then
+     echo 'Directory snpguest exists. Removing it...'
+     rm -rf snpguest
+  else
+     echo 'Directory snpguest does not exist.'
+  fi
+"
+
+  # Clone the repository and build the project
+  ssh_guest_command "git clone https://github.com/virtee/snpguest.git"
+  ssh_guest_command "cd snpguest  &&  source \$HOME/.cargo/env &&  cargo build --release"
+}
```
We already have logic for this, we decided we don't want to build the attestation tooling inside of the guest. We build it in the host and then scp it over to the guest. Look back at the snp guest attestation functions and build off those.

ramagali24 replied:
The SVSM binary cannot be built on Ubuntu 22.04, which uses binutils version 2.35. SVSM requires binutils version 2.39 or higher (as documented in the Coconut-SVSM installation procedure), which is available on Ubuntu 24.04, the OS most commonly used by the community. so i have to build snpguest directly on guest VM and my function does same.

ramagali24 replied:
I built the musl release of the snpguest  tool, updated the code, and verified it. The library compatibility issue is now resolved, and everything is working as expected and also removed setup_svsm_guest_attestation()

DGonzalezVillal commented on tools/snp.sh:
```diff
@@ -1107,16 +1152,102 @@ attest_guest() {
+  cd "$SETUP_WORKING_DIR"/AMDSEV/svsm
+
+  # Run the make command and save the output to a file
+  FW_FILE="$FW_FILE" make &> "$OUTPUT_FILE"
+
+  # Initialize the variable for Launch Digest
+  launch_digest=""
+
+  # Read the output file line by line to find the Launch Digest
+  while IFS= read -r line; do
+    if [[ "$line" == *"Launch Digest:"* ]]; then
+      launch_digest=$(echo "$line" | awk '{print $3}')
+      break
+    fi
+  done < "$OUTPUT_FILE"
```
I think all of this logic could be reduced.

You're running the make command, which I'm assuming at one point it uses igvmmeasure to calculate the measurement, but it also creates a lot of extra stuff that you put into your $OUTPUT_FILE. and then you read through this output file to look for the expected measurement. 

Can you not simply call igvmmeasure instead of running the whole make command? Idk what else is being created with this make command? IS there anything else you need in the $OUTPUT_FILE that is generated?

ramagali24 replied:
Yes, I will check. The igvmmeasure tool is an integral part of the SVSM compilation process. I will see if I can run it individually. I will update verify measurement process.

DGonzalezVillal commented on tools/snp.sh:
```diff
@@ -1148,6 +1279,10 @@ main() {
         shift; shift
         ;;
+      -s|--svsm)
+        SVSM=false
```
This is why all your logic seems to be backwards. If people are passing the svsm flag why are you making the variable false?

ramagali24 replied:
Yes, you got it finally.

DGonzalezVillal replied:
it shouldn't be that way, its confusing...

larrydewey replied:
I agree. This needs to switch, otherwise the logic is backward and doesn't make any sense. If the flag is passed, then the flag is enabled, not disabled.

DGonzalezVillal reviewed (changes requested):
Good fixes! This is what we were looking for per our previous discussions, just a couple more things to adjust.

Please fix the boolean logic for the svsm flag.

Also your verify measurement is adding repetitive code.

You will also need to rebase again because there were some prs merged yesterday, your code is behind.

ramagali24 commented:
To ensure that SNP is given priority over SVSM and that any SVSM-specific code changes fall back to the else condition, .
 -s|--svsm)
        SVSM=false

@larrydewey , I rebased my fork branch and submitted the commit earlier. Now, I have rebased my fork's main branch and then rebased and submitted the changes. This resolves most of your suggestions.

DGonzalezVillal commented:
Every time you are doing SNP workflows you are checking to see if SVSM is True, which in itself is confusing because it reads like SVSM is "enabled" let met do the SNP only workflows. This doesn't "give priority to SNP".

If SVSM is set to true you can do whatever small workflow you need for an svsm and if its false you can default to SNP. It reads better and it will make it easier to consolidate your logic. 

Or at least that's how I see it. @larrydewey  @ryansavino  idk what you guys think about that.

ramagali24 commented:
Absolutely, exploring different ways to communicate is beneficial. If you have any specific suggestions or need further assistance with the wording, please let me know.

The final goal is to avoid disrupting the SNP execution flow with unnecessary complications. We should prioritize the SNP code in the main section and place the SVSM code in the else part, due to slight differences in the processes.

ramagali24 commented:
I am thinking of adding SVSM="${SVSM:-false}", or using the function below:
is_svsm_true() {
  [ "$SVSM" = "true" ]
}
This makes the code more readable whenever there's a check for SVSM, and it always falls back to the ELSE loop if the condition is not met. Let me know your suggestions.

larrydewey commented on tools/snp.sh:
```diff
@@ -116,6 +118,7 @@ usage() {
   >&2 echo "    stop-guests           Stop all SNP guests started by this script"
   >&2 echo "  where OPTIONS are:"
   >&2 echo "    -n|--non-upm          Build AMDSEV non UPM kernel (sev-snp-devel)"
+  >&2 echo "    -s|--svsm              Build coconut-svsm components, launch guest and verity attestaion & measurement"
```
```suggestion
  >&2 echo "    -s|--svsm             Build coconut-svsm components, launch guest and verity attesation & measurement"
```

larrydewey commented on tools/snp.sh:
```diff
@@ -1073,50 +1100,140 @@ generate_snp_expected_measurement() {
+      local expected_measurement=$(generate_snp_expected_measurement)
+      echo -e "\nExpected Measurement (sev-snp-measure): ${expected_measurement}"
+    else
+      local expected_measurement=$(generate_svm_expected_measurement)
```
Should we name the function `generate_svsm_expected_measurement` instead of `generate_svm_expected_measurement`

larrydewey commented on tools/snp.sh:
```diff
@@ -525,22 +525,17 @@ save_binary_paths() {
-IGVM_FILE="${SETUP_WORKING_DIR}/AMDSEV/svsm/bin/coconut-qemu.igvm"
 EOF
-fi
+  if [ "$SVSM" = true ]; then
```
Logic is still reversed here. This doesn't make logical sense.

larrydewey commented on tools/snp.sh:
```diff
@@ -563,19 +558,16 @@ copy_launch_binaries() {
 KERNEL_BIN="${LAUNCH_WORKING_DIR}/$(basename "${KERNEL_BIN}")"
 EOF
-fi
+  if [ "$SVSM" = true ]; then
```
Logic is backward here. Please fix

larrydewey commented on tools/snp.sh:
```diff
@@ -671,16 +663,12 @@ build_and_install_amdsev() {
-       
-    fi	
-		  
+  if [ "$SVSM" = true ]; then
```
Logic backward. Please Fix.

larrydewey commented on tools/snp.sh:
```diff
@@ -671,16 +663,12 @@ build_and_install_amdsev() {
+  else
+    AMDSEV_URL=${AMDSEV_SVSM_URL}
+    amdsev_branch="svsm-latest"
+	fi			  
```
Fix your indentation here.

larrydewey commented on tools/snp.sh:
```diff
@@ -700,7 +688,7 @@ build_and_install_amdsev() {
   if [ "$SVSM" = true ]; then
```
Logic backward. Please Fix.

larrydewey replied:
No. You can't skip this for the sake of speeding up the merging of your PR. Fixing the logic for this does nothing to break any level of the build process. It literally is centered around a conditional change. If you don't do clean up, and you were to build the SVSM then build the standard image, it will break the image as it doesn't clean up the SVSM from the previous build. You need to address it and fix it.

larrydewey commented on tools/snp.sh:
```diff
@@ -1112,138 +1100,136 @@ generate_snp_expected_measurement() {
-  # Verifies that ARK, ASK and VCEK are all properly signed
-  ssh_guest_command "./$Home/snpguest/target/release/snpguest verify certs ."
+    # Request and display the snp attestation report with random data
+    if [ "$SVSM" = true ]; then
+      ssh_guest_command "sudo ./snpguest report attestation-report.bin request-data.txt --random"
+    else
+      ssh_guest_command "sudo ./snpguest report attestation-report.bin request-data.txt --random --vmpl 3"
+    fi
```
Instead of re-writing the entire call, just create the string without the VMPL. Add your conditional logic below, appending the vmpl if SVSM is enabled...

larrydewey commented on tools/snp.sh:
```diff
@@ -1112,138 +1100,136 @@ generate_snp_expected_measurement() {
-    echo 'Installing Rust...'
-    source ""${HOME}"/.cargo/env" 2>/dev/null || true
+    # Use appropriate measurement function based on SVSM
+    if [ "$SVSM" = true ]; then
```
Logic backward. Please Fix.

larrydewey commented on tools/snp.sh:
```diff
@@ -1112,138 +1100,136 @@ generate_snp_expected_measurement() {
+# Copy the built binary to the home directory
+cp target/release/snpguest "$HOME/"
+EOF
+}
```
This entire function can be removed. The program is already building the tool once. It doesn't make sense to duplicate this again.

larrydewey commented on tools/snp.sh:
```diff
@@ -1112,138 +1100,136 @@ generate_snp_expected_measurement() {
-    if [[ "$line" == *"Launch Digest:"* ]]; then
-      launch_digest=$(echo "$line" | awk '{print $3}')
-      break
+generate_svm_expected_measurement() {
```
Rename to be `svsm` for consistency sake

ryansavino commented:
There's a lot of comments here that need to be addressed. Please try to clean things up, and then I'll a look at the rest of the PR to provide my review.

larrydewey commented on tools/snp.sh:
```diff
@@ -646,9 +663,16 @@ set_acl_for_sev_device() {
+  
+  if [ "$SVSM" = true ]; then
+    AMDSEV_URL=${AMDSEV_SVSM_URL}
+    amdsev_branch="svsm-latest-fixes"    	 
```
You should probably use the global value here, as you already defined this string at the top?

larrydewey commented on tools/snp.sh:
```diff
@@ -664,10 +688,11 @@ build_and_install_amdsev() {
   git fetch current "${amdsev_branch}"
   git checkout "current/${amdsev_branch}"
-
+ 
```
Unnecessary Spacing

larrydewey commented on tools/snp.sh:
```diff
@@ -664,10 +688,11 @@ build_and_install_amdsev() {
   # Based on latest AMDSEV documentation
   # Delete the ovmf/ directory prior to the build step for ovmf re-initialization
   [ ! -d "ovmf" ] || rm -rf "ovmf"
+ 
```
Unnecessary Spacing

larrydewey commented on tools/snp.sh:
```diff
@@ -1073,48 +1103,130 @@ generate_snp_expected_measurement() {
+echo 'Installing necessary packages...'
+sudo apt-get install -y git build-essential libtss2-dev tpm2-tools
+
+# Install Rust
+echo 'Installing Rust...'
+source "$HOME/.cargo/env" 2>/dev/null || true
+curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y
+source "$HOME/.cargo/env" 2>/dev/null
+
+# Create the 'tools' directory if it does not exist
+if [ ! -d "tools" ]; then
+    echo 'Directory tools does not exist. Creating it...'
+    mkdir tools
+else
+    echo 'Directory tools already exists.'
+fi
+
+# Change to 'tools' directory
+cd tools || { echo 'Failed to change directory to tools'; exit 1; }
+
+# Check if 'snpguest' directory exists and remove it if it does
+if [ -d "snpguest" ]; then
+    echo 'Directory snpguest exists. Removing it...'
+    rm -rf snpguest
+else
+    echo 'Directory snpguest does not exist.'
+fi
+
+# Clone the repository into the 'tools' directory
+git clone https://github.com/virtee/snpguest.git
+
+# Build the project inside the 'snpguest' directory
+cd snpguest || { echo 'Failed to change directory to snpguest'; exit 1; }
+source "$HOME/.cargo/env"
+cargo build --release
+
+# Copy the built binary to the home directory
+cp target/release/snpguest "$HOME/"
+EOF
 }
```
Still need to remove this function. This is already done in another section of the code.

ramagali24 replied:
My host is running Ubuntu 24.02, which has glibc version 2.39. If you build snpguest on this host, it won't run on an Ubuntu 22.04 guest VM because the guest uses glibc version 2.35. Therefore, I need to write a new function to build snpguest directly on the guest VM. (Error: ./snpguest: /lib/x86_64-linux-gnu/libc.so.6: version 'GLIBC_2.38' not found, required by ./snpguest). The existing function works if both the host and guest are running the same version of Ubuntu.

ramagali24 replied:
The SVSM binary cannot be built on Ubuntu 22.04, which uses binutils version 2.35. SVSM requires binutils version 2.39 or higher (as documented in the Coconut-SVSM installation procedure), which is available on Ubuntu 24.04, the OS most commonly used by the community.

larrydewey replied:
> My host is running Ubuntu 24.02, which has glibc version 2.39. If you build snpguest on this host, it won't run on an Ubuntu 22.04 guest VM because the guest uses glibc version 2.35. Therefore, I need to write a new function to build snpguest directly on the guest VM. (Error: ./snpguest: /lib/x86_64-linux-gnu/libc.so.6: version 'GLIBC_2.38' not found, required by ./snpguest). The existing function works if both the host and guest are running the same version of Ubuntu.

If you are running into library compatibility issues, just build the musl release of the tool. This should fix the library issue.

ramagali24 replied:
Thank you for the suggestion. I built the musl release of the snpguest  tool, updated the code, and verified it. The library compatibility issue is now resolved, and everything is working as expected and also removed setup_svsm_guest_attestation()

larrydewey commented on tools/snp.sh:
```diff
@@ -116,6 +118,7 @@ usage() {
   >&2 echo "    stop-guests           Stop all SNP guests started by this script"
   >&2 echo "  where OPTIONS are:"
   >&2 echo "    -n|--non-upm          Build AMDSEV non UPM kernel (sev-snp-devel)"
+  >&2 echo "    -s|--svsm              Build coconut-svsm components, launch guest and verify attestaion & measurement"
```
```suggestion
  >&2 echo "    -s|--svsm             Build coconut-svsm components, launch guest and verify attestation & measurement"
```

larrydewey commented on tools/snp.sh:
```diff
@@ -524,10 +527,14 @@ save_binary_paths() {
 INITRD_BIN="${GENERATED_INITRD_BIN}"
 KERNEL_BIN="${guest_kernel}"
 EOF
+  if [ "$SVSM" = true ]; then
```
Is there any reason why we don't just do:

```suggestion
  if "$SVSM"; then
```

ramagali24 replied:
Agreed.

larrydewey commented on tools/snp.sh:
```diff
@@ -1225,7 +1349,7 @@ main() {
       sudo modprobe -r kvm_amd
       sudo modprobe kvm_amd debug_swap=0
-
+      
```
Extra tab?

DGonzalezVillal commented on tools/snp.sh:
```diff
@@ -646,9 +661,16 @@ set_acl_for_sev_device() {
+    amdsev_branch="${1:-${AMDSEV_SVSM_BRANCH}}"    	 
+  else
+    local amdsev_branch="${1:-${AMDSEV_DEFAULT_BRANCH}}"
+  fi			    
```
After looking at the logic of when this function is called ([line 1325](https://github.com/amd/sev-utils/pull/19/files#diff-53302c0f5bf2a564b2d0cbdd4e23d1750f4257869931c200697947ae6ec86143R1325)) you can move this logic there. Then you wouldn't really have to modify this function, since this is really the only change you did here.

ramagali24 replied:
Moved this logic to main function. Please check.

pegahnikbakht commented:
Does this pull request prepare both host and guest kernel for svsm support or only it prepares the host kernel?

ramagali24 commented:
> Does this pull request prepare both host and guest kernel for svsm support or only it prepares the host kernel?

Yes, it prepares both host and guest kernel build's of svsm.

larrydewey commented on tools/snp.sh:
```diff
@@ -116,6 +119,7 @@ usage() {
   >&2 echo "    stop-guests           Stop all SNP guests started by this script"
   >&2 echo "  where OPTIONS are:"
   >&2 echo "    -n|--non-upm          Build AMDSEV non UPM kernel (sev-snp-devel)"
+  >&2 echo "    -s|--svsm              Build coconut-svsm components, launch guest and verify attestation & measurement"
```
Extra space
```suggestion
  >&2 echo "    -s|--svsm             Build coconut-svsm components, launch guest and verify attestation & measurement"
```

larrydewey commented on tools/snp.sh:
```diff
@@ -199,6 +203,11 @@ install_rust() {
 
   # Install rust
   curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs -sSf | sh -s -- -y
+  #Beeded to build Cococnut-SVSM
+  if "$SVSM"; then
+    rustup target add x86_64-unknown-linux-musl
+    cargo install cbindgen
+  fi
```
```suggestion
  # Needed to build Cococnut-SVSM
  if "$SVSM"; then
    rustup target add x86_64-unknown-linux-musl
    cargo install cbindgen
  fi
```

@ryansavino it may make sense to just build the musl for the entire project (fixing the same issue with backward compatibility with older distros, etc. Thoughts?

larrydewey commented on tools/snp.sh:
```diff
@@ -275,6 +284,13 @@ install_dependencies() {
   # Needed to build 6.11.0-rc3 SNP kernel on the host
   pip install tomli
   
+  #Needed to build Coconut-SVSM
```
```suggestion
  # Needed to build Coconut-SVSM
```

larrydewey commented on tools/snp.sh:
```diff
@@ -275,6 +284,13 @@ install_dependencies() {
   pip install tomli
   
+  #Needed to build Coconut-SVSM
+  if [ "$SVSM" ]; then
+    sudo apt install -y libcunit1 libcunit1-dev build-essential libclang-dev autoconf \
+      autoconf-archive pkg-config automake libssl-dev libc6-dev gcc-multilib binutils make musl musl-tools
+  fi
```
@ramagali24 Definitely want to remove the brackets here.
@ryansavino same thoughts about this part too. Should we look at integrating this 
```suggestion
  if $SVSM; then
    sudo apt install -y libcunit1 libcunit1-dev build-essential libclang-dev autoconf \
      autoconf-archive pkg-config automake libssl-dev libc6-dev gcc-multilib binutils make musl musl-tools
  fi
```

larrydewey commented on tools/snp.sh:
```diff
@@ -786,13 +815,21 @@ setup_and_launch_guest() {
-  add_qemu_cmdline_opts "-initrd ${INITRD_BIN}"
-  add_qemu_cmdline_opts "-kernel ${KERNEL_BIN}"
-  add_qemu_cmdline_opts "-append \"${GUEST_KERNEL_APPEND}\""
+  if "$SVSM"; then
+    add_qemu_cmdline_opts "-initrd ${INITRD_BIN}"
+    add_qemu_cmdline_opts "-kernel ${KERNEL_BIN}"
+    add_qemu_cmdline_opts "-append \"${GUEST_KERNEL_APPEND}\""
+  else
+    add_qemu_cmdline_opts "-bios ${OVMF_BIN}"
+  fi
```
This seems backward to me?

ramagali24 replied:
Thanks. The changes have been corrected now.

larrydewey commented on tools/snp.sh:
```diff
@@ -1026,7 +1068,7 @@ get_cpu_code_name() {
 }
 
-generate_snp_expected_measurement() {
+generate_snp_expected_measurement() {  
```
extra space

larrydewey commented on tools/snp.sh:
```diff
@@ -1073,50 +1115,89 @@ generate_snp_expected_measurement() {
+    fi
+    # Convert svsm_measurement to lowercase
+    svsm_measurement=$(echo "$svsm_measurement" | tr '[:upper:]' '[:lower:]')
+    echo ${svsm_measurement}
```
Remove debug statement

larrydewey commented on tools/snp.sh:
```diff
@@ -1240,7 +1328,7 @@ main() {
       wait_and_retry_command verify_snp_guest
       setup_guest_attestation
-      attest_guest
+      attest_guest  	  
```
Remove extra space
```suggestion
      attest_guest
```

Commit:
chnages made based on latest reviews

larrydewey commented on tools/snp.sh:
```diff
@@ -786,13 +814,20 @@ setup_and_launch_guest() {
   add_qemu_cmdline_opts "-initrd ${INITRD_BIN}"
   add_qemu_cmdline_opts "-kernel ${KERNEL_BIN}"
   add_qemu_cmdline_opts "-append \"${GUEST_KERNEL_APPEND}\""
+  if [ -z "$SVSM" ]; then
+    add_qemu_cmdline_opts "-bios ${OVMF_BIN}"
+  fi
```
Now that you are hard-coding the value to false at the top of the function, this won't work anymore, as the variable has been set with a value.

ramagali24 replied:
Please check once.

larrydewey commented on tools/snp.sh:
```diff
@@ -786,13 +814,18 @@ setup_and_launch_guest() {
   add_qemu_cmdline_opts "-initrd ${INITRD_BIN}"
   add_qemu_cmdline_opts "-kernel ${KERNEL_BIN}"
   add_qemu_cmdline_opts "-append \"${GUEST_KERNEL_APPEND}\""
+  [[ "$SVSM" != true ]] && add_qemu_cmdline_opts "-bios ${OVMF_BIN}"
```
Not a blocker for me, but you could also just negate the boolean value:

```bash
if ! "$SVSM"; then
   # Stuff here...
fi

larrydewey reviewed (dismissed):
I think this addressed my concerns. @ryansavino and @DGonzalezVillal What are your thoughts?

pegahnikbakht commented:
I tried the PR and this what I get, after setting up the host and doing reboot: 
```
sudo dmesg | grep SEV
[    0.000000] SEV-SNP: RMP table physical range [0x0000000097f00000 - 0x00000000a84fffff]
[   15.971036] ccp 0000:47:00.1: SEV firmware update successful
[   16.957109] ccp 0000:47:00.1: SEV API:1.55 build:17
[   16.968043] ccp 0000:47:00.1: SEV-SNP API:1.55 build:17
[   17.001398] kvm_amd: SEV enabled (ASIDs 100 - 509)
[   17.001400] kvm_amd: SEV-ES enabled (ASIDs 1 - 99)
[   17.001402] kvm_amd: SEV-SNP enabled (ASIDs 1 - 99)
[   55.694404] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[   69.696023] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[   83.714032] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[   98.741790] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[  112.706735] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[  126.726123] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[  138.749160] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[  151.755308] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[  162.736310] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[  177.706665] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[  189.709388] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[  201.727346] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[  216.711789] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[  229.729728] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[  254.713263] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[  270.741829] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[  286.746275] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[  300.714161] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[  315.665391] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[  326.740066] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[  340.334649] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[  352.739657] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[  364.751004] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[  377.723099] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[  391.715254] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[  404.725178] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[  418.724743] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[  429.734649] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[  444.723452] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
[  456.752758] kvm_amd: SEV-SNP requires private memory support via guest_memfd.
```

Same error I get when setting up the SVSM directly from coconut svsm github repo.

ramagali24 commented:
@pegahnikbakht , Could you please open a new issue so we can discuss and provide suggestions for your setup problems? I don't see the same issue you're experiencing with the PR; it seems more related to your setup and installation

DGonzalezVillal commented on tools/snp.sh:
```diff
@@ -116,6 +119,7 @@ usage() {
   >&2 echo "    stop-guests           Stop all SNP guests started by this script"
   >&2 echo "  where OPTIONS are:"
   >&2 echo "    -n|--non-upm          Build AMDSEV non UPM kernel (sev-snp-devel)"
+  >&2 echo "    -s|--svsm             Build coconut-svsm components, launch guest and verify attestation & measurement"
```
```suggestion
  >&2 echo "    -s|--svsm             Build, launch and attest coconut-svsm"
```

DGonzalezVillal commented on tools/snp.sh:
```diff
@@ -275,6 +284,12 @@ install_dependencies() {
   pip install tomli
   
+  # Needed to build Coconut-SVSM
+  if $SVSM; then
+    sudo apt install -y libcunit1 libcunit1-dev build-essential libclang-dev autoconf \
+      autoconf-archive pkg-config automake libssl-dev libc6-dev gcc-multilib binutils make musl musl-tools
+  fi
```
Some of these are already being installed (build-essential, libssl-dev, pkg-config)

DGonzalezVillal commented on tools/snp.sh:
```diff
@@ -910,9 +943,14 @@ setup_guest_attestation() {
-  cargo build -r
-  scp_guest_command target/release/snpguest "${GUEST_USER}@localhost:/home/${GUEST_USER}"
+  
+  if "$SVSM"; then
+    cargo build --release --target x86_64-unknown-linux-musl
+    scp_guest_command target/x86_64-unknown-linux-musl/release/snpguest "${GUEST_USER}@localhost:/home/${GUEST_USER}"
+  else
+    cargo build -r
+    scp_guest_command target/release/snpguest "${GUEST_USER}@localhost:/home/${GUEST_USER}"
+  fi
```
@larrydewey & @ryansavino Not the biggest deal, but Rama had to add this due to dependency differences in Ubuntu 22.04 and 24.04. Do we want to default to musl builds? If I'm not mistaken this will happen to anyone that has a 24.04 host that tries to use the script to launch a 22.04 guest. It's unrelated to svsm.

larrydewey replied:
I agree. We should probably be using the musl binary when building for something like this to remove failure layers where possible.

DGonzalezVillal reviewed (commented):
Only a couple of suggestions and one general question, but overall I'm good with how it looks now!

DGonzalezVillal commented:
Oh also I just thought of this, but maybe update the snp.md documentation with updated instructions to use the SVSM flag. Or do you want that to be a separate PR @ryansavino

Commit:
Updated snp.md and review suggestions

ramagali24 commented:
> Oh also I just thought of this, but maybe update the snp.md documentation with updated instructions to use the SVSM flag. Or do you want that to be a separate PR @ryansavino

Updated snp.md as well.

DGonzalezVillal commented on docs/snp.md:
````diff
@@ -56,6 +56,10 @@ Setup the host by building SNP patched versions of qemu, ovmf and the linux kern
 ```
 ./snp.sh setup-host
 ```
+Coconut-SVSM: To set up the host by building IGVM, SVSM, QEMU, OVMF, and the Linux kernel, use the following command:
+```
+./snp.sh --svsm setup-host
+```
````
```suggestion
The `--svsm` option can be specified with the above command to set-up the host for SVSM. This will build a Coconut-SVSM kernel as well as the IGVM, SVSM, OVMF and QEMU launch dependencies.
```
```

ramagali24 replied:
You are asking remove command and add verbal explanation. I would like to keep command as well in documentation. This gives clear instruction on where they should add "--svsm" option. @ryansavino, any suggestions?

DGonzalezVillal commented on docs/snp.md:
````diff
@@ -73,6 +77,10 @@ the following command:
 ```
 ./snp.sh launch-guest
 ```
+On the Coconut-SVSM setup, use the following command:
+```
+./snp.sh --svsm launch-guest
+```
````
```suggestion
The `--svsm` option can be specified in the above command to launch a Coconut-SVSM guest. The `setup-host` command must be run with this same option in order to have the necessary dependencies to launch an SVSM guest.
```

ramagali24 replied:
You are asking remove command and add verbal explanation. I would like to keep command as well in documentation. This gives clear instruction on where they should add "--svsm" option. @ryansavino, any suggestions?

DGonzalezVillal commented on docs/snp.md:
````diff
@@ -86,6 +94,10 @@ Attest the guest using the following command:
 ```
 ./snp.sh attest-guest
 ```
+On the Coconut-SVSM setup, use the following command:
+```
+./snp.sh --svsm attest-guest
+```
````
```suggestion
Use the `--svsm` option in the above command to attest a Coconut-SVSM guest.
```

ramagali24 replied:
You are asking remove command and add verbal explanation. I would like to keep command as well in documentation. This gives clear instruction on where they should add "--svsm" option. @ryansavino, any suggestions?

DGonzalezVillal commented on docs/snp.md:
```diff
@@ -96,6 +108,8 @@ is then checked and verified against the launch measurement that is output from
 [virtee/snpguest](https://github.com/virtee/snpguest) tool. If the two measurements 
 match, then the test returns with a successful output.
 
+Note: On the Coconut-SVSM setup, the igvmmeasure tool (part of the SVSM build) is used to calculate the expected launch measurement.
+
```
```suggestion
In the case of Coconut-SVSM attestation, the [igvmmeasure](https://github.com/coconut-svsm/svsm/tree/main/igvmmeasure) tool is used, which will only measure the IGVM file used to launch the guest. This expected measurement should match with the measurement on the attestation report of the SVSM guest.
```

DGonzalezVillal reviewed (changes requested):
I'm good with the code, just some nits on the readme to keep the format consistent. But overall you are pretty much done!

Commit:
updated snp.md with reviewer suggestions

Commit:
updated snp.md with reviewer suggestion

DGonzalezVillal commented:
Address musl build comment I brought up, Squash your commits and then we should be good!

Also resolve conversations of topics you've fixed already.

pegahnikbakht commented:
What happened to this PR? no more existence?
